File: api/src/org/ufomatics/archipelago/model/MetaModelImpl.java
package org.ufomatics.archipelago.model;

import org.ufomatics.archipelago.model.MetaModel;
import org.ufomatics.archipelago.model.MetaType;
import org.ufomatics.archipelago.model.MetaFunction;

import java.util.Set;
import java.util.HashSet;
import java.io.Serializable;

/**
 * @author Jérôme Beau
 * @version 8 mai 2006 01:53:57
 */
public class MetaModelImpl implements MetaModel, Serializable {
    private Set<MetaType> classes = new HashSet<MetaType>();
    private Set<MetaFunction> functions = new HashSet<MetaFunction>();

    public MetaModelImpl() {
        addMetaType(MetaType.TEXT);
        addMetaType(MetaType.NUMBER);
        addMetaType(MetaType.IMAGE);
    }

    public Set<MetaType> getClasses() {
        return classes;
    }

    public void setClasses(Set<MetaType> classes) {
        this.classes = classes;
    }

    public boolean isEmpty() {
        return classes.isEmpty() && functions.isEmpty();
    }

    public void addMetaType(MetaType metaType) {
        classes.add (metaType);
    }

    public void addFunction(MetaFunction metaFunction) {
        functions.add (metaFunction);
    }

    public Set<MetaFunction> getFunctions() {
        return functions;
    }

    public void setFunctions(Set<MetaFunction> functions) {
        this.functions = functions;
    }

    public MetaType createType() {
        return new MetaTypeImpl();
    }

    public MetaFunction createFunction() {
        return new MetaFunctionImpl();
    }

    public MetaType getType (String typeName) {
        for (MetaType aClass : classes) {
            if (aClass.getName().equals (typeName)) {
                return aClass;
            }
        }
        return null;
    }
}

File: src/archipelago/api/model/MetaModelImpl.ts
import {MetaModel} from "archipelago/api/model/MetaModel"
import {MetaFunction} from "archipelago/api/model/MetaFunction"
import {IMAGE, MetaType, NUMBER, TEXT} from "archipelago/api/model/MetaType"
import {MetaTypeImpl} from "archipelago/api/model/MetaTypeImpl"
import {MetaFunctionImpl} from "archipelago/api/model/MetaFunctionImpl"
import {HashSet, JSerializable, JSet} from "ts-jsdk"

/**
 *
 */
export class MetaModelImpl implements MetaModel, JSerializable {
  private classes: JSet<MetaType> = new HashSet<MetaType>()
  private functions = new HashSet<MetaFunction>()

  constructor() {
    this.addMetaType(TEXT)
    this.addMetaType(NUMBER)
    this.addMetaType(IMAGE)
  }

  getClasses(): JSet<MetaType> {
    return this.classes
  }

  setClasses(classes: JSet<MetaType>) {
    this.classes = classes
  }

  isEmpty(): boolean {
    return this.classes.isEmpty() && this.functions.isEmpty()
  }

  addMetaType(metaType: MetaType): void {
    this.classes.add(metaType)
  }

  addFunction(metaFunction: MetaFunction): void {
    this.functions.add(metaFunction)
  }

  getFunctions(): JSet<MetaFunction> {
    return this.functions
  }

  setFunctions(functions: HashSet<MetaFunction>) {
    this.functions = functions
  }

  createType(name: string): MetaType {
    return new MetaTypeImpl(name)
  }

  createFunction(): MetaFunction {
    return new MetaFunctionImpl()
  }

  getType(typeName: string): MetaType | null {
    for (let aClass of this.classes) {
      if (aClass.getName() === typeName) {
        return aClass
      }
    }
    return null
  }
}
